Save 04/03/2025

eoin2406 opened the pull request:
I created the base game. This includes the player's HP, monster's HP, a healing item, inventory, room description, contents. Player cannot receive damage from the monster. The game ends when the monster is killed.

eoin2406 committed:
Save 04/03/2025

Created the game, including player health, minotaur health, healing, inventory, room description, contents. Player cannot receive damage from the minotaur. Game ends when minotaur is killed.

notxcharles commented on Room.cs:
```diff
@@ -1,17 +1,25 @@
         {
             return description;
         }
+        public string RoomContents()
```
You could rename this to GetRoomContents(), this would make the code more readable and self-explanatory

BenSt-Clair replied:
foreach (x in inventory){console.WriteLine(x.Name)} - for searching through inventory and returning all possible objects. you need to think of this room class as not being for one room but for all possible rooms. otherwise you may as well just code for a single object, or a particular room, unless all your rooms only contain an old map.

eoin2406 replied:
> foreach (x in inventory){console.WriteLine(x.Name)} - for searching through inventory and returning all possible objects. you need to think of this room class as not being for one room but for all possible rooms. otherwise you may as well just code for a single object, or a particular room, unless all your rooms only contain an old map.

I changed this to use "inventory[0]" after following Charles' suggestion, although I do feel like that might be more useful if more than one room is to be added.
Currently there is only one item in the game, this being the old map, so it should not be necessary. I will keep these points in mind though.
Thank you, Ben.

notxcharles commented on Player.cs:
```diff
@@ -15,11 +16,32 @@ public Player(string name, int health)
         {
             return string.Join(", ", inventory);
         }
+        public int getHealth()
```
Capitalise (Pascal Case) getHealth() to GetHealth(). This aligns better with the C# naming conventions and aligns with how other functions have been named

notxcharles commented on Room.cs:
```diff
@@ -1,17 +1,25 @@
         }
+        public string RoomContents()
+        {
+            return ("an old map");
```
This wouldn't change if the item in our inventory changed. In case it does, `return ("an old map");` could be changed to something like:
```cs
string item = inventory[0];
return item;
```

notxcharles commented on Monster.cs:
```diff
@@ -0,0 +1,28 @@
+    internal class Monster
+    {
+        private int Health { get; set; }
+        private string Name { get; set; }
```
Good use of Encapsulation

notxcharles commented on Monster.cs:
```diff
@@ -0,0 +1,28 @@
+            this.Name = Name;
+            this.Health = Health;
+        }
+        public int getHealth()
```
Similar to a previous comment I made, this function should be renamed to use Pascal Case -> `GetHealth()`

notxcharles commented on Monster.cs:
```diff
@@ -0,0 +1,28 @@
+        {
+            return this.Health;
+        }
+        public string attackDmg()
```
This function could be renamed to something easier to understand, perhaps something like GetAttackDamage()

BenSt-Clair replied:
Hey, Eoin. This looks good, but right now that console line states specifically for a minotaur where you could use interpolation to input Monster.Name, thereby making it general for all possible monsters and not just one instantiation or specific monster.

eoin2406 replied:
> Hey, Eoin. This looks good, but right now that console line states specifically for a minotaur where you could use interpolation to input Monster.Name, thereby making it general for all possible monsters and not just one instantiation or specific monster.

Hi Ben, similar to the last point, at the moment there is only one monster in the game, so I feel like this won't be required. However it is a brilliant point that you have raised, and I will definitely act upon that if I am to add another monster to the game.
Thank you very much.

notxcharles commented on Game.cs:
```diff
@@ -1,26 +1,83 @@
+            if (player.CheckForItems() == true)
             {
-                // Code your playing logic here
+                Console.WriteLine("Type i to view the inventory.");
```
I like the logic here where the player is not given the option to check his inventory if it is empty

notxcharles commented on Game.cs:
```diff
@@ -1,26 +1,83 @@
+                    }
+                    else
+                    {
+                        Console.WriteLine("Invalid input.");
```
Good use of error catching. If the player doesn't input the correct character then the program tells them so.

BenSt-Clair replied:
Have you considered making an item class instead of just having items as strings?

eoin2406 replied:
> Have you considered making an item class instead of just having items as strings?

Currently there is only the one item, however my previous point still stands and I will definitely act upon this idea in the future if I am to add more items.
Thank you.

notxcharles commented on Player.cs:
```diff
@@ -15,11 +16,32 @@ public Player(string name, int health)
-
+        inventory.Add(item);
         }
         public string InventoryContents()
```
InventoryContents() could be renamed to GetInventoryContents()

notxcharles reviewed (changes requested):
**Overall** a good project. I like the way that you have used Encapsulation and Abstraction to protect some variables from being used outside of the class. You've also provided getters and setters for any of the class properties that may be accessed from outside of the class. The logic which automatically stops the game when the monster has 0 health is very nice. Each class and function just doing one thing was nice to see (Single Responsibility Principle)

**Additional Feedback:**
I think that some of the functions could be renamed to either enhance readability or make it more self explanatory. 
There is a potential logic error in Room.cs that I have commented on, it would be nice for this to be fixed before further adding to the game.
I would like to see the monster fight back and do damage to the player. At the moment the healing functionality doesn't really need to be present

**Requirements:**
1. Classes and Objects:
- [x] Game.cs file handles the game flow and initialises the player and one simple room
	- All the logic for the game is present. I can choose to heal my character, attack the monster, pickup an item and view my inventory.
	- There has been an additional class for Monster created. This is nice and shows a difference between the player and any monsters

- [x] Player.cs tracks the player's name and a signle attribute, such as health or a basic inventory
	- Player has a name, health value and inventory. Name and Health have a getter and a setter (private setter for Name and Health).
	- Clear method for adding to the inventory (PickUpItem()) and getting a list of the inventory's contents (InventoryContents()).
	- Method for healing the character (Heal()) that can be called outside of the class
- [x] Room.cs represents a single room in the game with a description and possibly an item
	- Room has a description and contains an item. Method for getting the description and also for getting the item in the room

2. Encapsulation and Abstraction
- [x] Use private fields and provide get/set methods where necessary
	- Used within Player.cs and Room.cs. Getters and setters for multiple class properties

3. Methods & Constructors
- [x] Create constructors for the Player and Room classes to initialise their attributes.
	- Player, Room, Game and Monster constructors

	Implement basic methods:
	- [x] Player.PickUpItem(): Adds a single item to the player's inventory.
		- Picking up an item adds the item to the players inventory. The PickUpItem() even takes a parameter item so in the future the item could change.
	- [x] Room.GetDescription(): Returns the description of
		- Room.GetDescription() returns a description of the room

4. Basic Game Interaction
- [x] The player starts in a single room and can:
	- [x] View the room's description.
		- User is given a description of the room at the start of the round and game
	- [x] Display their current status (e.g., inventory and health).
		- Player.cs and monster.cs holds this information. At the start of each round, the user is told the player's health and the monster's health. The user is also informed of the damage done to the monster.

5. Error Checking:
- [x] Add simple error checking, such as preventing the player from entering a room that does not exist.
	- Validate and check that the correct user input is checked

BenSt-Clair commented:
Error Handling:
  Your code won't throw up exceptions and this is good. However you might want to consider using .ToLower() and Trim() functions for player input. These will ensure that if a player enters ' a' [note the white space] or 'A' to attack, then these inputs will also be accepted. You've used conditionals which is fine. I haven't seen any System.Diagnostics.Debug.Assert() used to check exceptions won't be thrown preemptively. And there are no try and catch statements. As you continue to develop your code, you might want to consider deploying these techniques. Try and catch are especially useful when your code gets more complex and you want to use a function that might only apply some of the time and throw exceptions on others. 

Encapsulation and Abstraction:
  You've used the internal scope and private fields which is great. I noticed that you might have some confusion as to what classes are for and how they are distinct from objects. This is because [and i mentioned above] your classes' functions often return stock responses to the console that isn't necessarily going to apply to every object your instantiate with said class. Simple interpolation is test way to get around this in most cases.
  There are no abstract classes from what I can tell and you have yet to deploy inheritance.

  Methods and Constructors:
  You've made the constructors well and differently from how we were taught in class which I found interesting. Each class for room returns a description and the status of the player can be found throughout the game. As for pickupitem... Don't you want to just pick up one item at a time from the room? Also why does the minotaur just stand there and take your attacks? What do you have against the noble and majestic creature that is the minotaur that you should so callously transform it into a punching bag to be pommelled until death? Also you might want a maximum health and a finite number of healing supplements otherwise I *am* going to play this game healing my character until i max out space on my computer to contain all the digits that'll be my player health.

eoin2406 committed:
Charles' Suggested Changes

For presentation purposes:
Changed the function: "Heal" to "HealPlayer"
Changed the function: "getHealth" to "GetHealth"
Changed the function: "RoomContents" to "GetRoomContents"
Changed the function: "attackDmg" to "GetAttackDmg"
Changed the function: "InventoryContents" to "GetInventoryContents"
Changed other function names to follow the Pascal Case format.

Using "Get" in these function names is nice as it shows that the function will display them, e.g: "GetHealth" displays the current health of the player, etc...

For presentation purposes:
Changed: "a" to "attack"
Changed: "h" to "heal"
Changed: "i" to "inv" (inventory)
Changed: "p" to "pick" (pickup item)

Although this makes typing the actions a longer process, it seems a worthwhile edit for presentation purposes. All of the descriptions in the interface are now neater, with a brief description of their purposes for the player.

Pascal Case is now used for every function name.
These changes were made for readability purposes, as well as to make the function names more consistent and clear.

Description attacking the minotaur has been edited, making it feel more appropriate.
Description for defeating the minotaur has been edited, making it feel more appropriate.
Minotaur HP has increased to 80, making it feel more challenging, as the player has 100HP.

Added a new function that causes the player to receive damage from the minotaur, making the heal function actually have a use.

After reading the reviews suggested by Charles, I acted upon the feedback and a lot of it was very helpful advice. I really liked the idea of changing the function names to begin with "Get" if they are to display their purposes on the game screen. It makes the code feel a lot tidier, complete and enhances readability. I added a function that causes the monster to attack the player to add a level of difficulty to the game.

eoin2406 commented:
> Error Handling: Your code won't throw up exceptions and this is good. However you might want to consider using .ToLower() and Trim() functions for player input. These will ensure that if a player enters ' a' [note the white space] or 'A' to attack, then these inputs will also be accepted. You've used conditionals which is fine. I haven't seen any System.Diagnostics.Debug.Assert() used to check exceptions won't be thrown preemptively. And there are no try and catch statements. As you continue to develop your code, you might want to consider deploying these techniques. Try and catch are especially useful when your code gets more complex and you want to use a function that might only apply some of the time and throw exceptions on others.
> 
> Encapsulation and Abstraction: You've used the internal scope and private fields which is great. I noticed that you might have some confusion as to what classes are for and how they are distinct from objects. This is because [and i mentioned above] your classes' functions often return stock responses to the console that isn't necessarily going to apply to every object your instantiate with said class. Simple interpolation is test way to get around this in most cases. There are no abstract classes from what I can tell and you have yet to deploy inheritance.
> 
> Methods and Constructors: You've made the constructors well and differently from how we were taught in class which I found interesting. Each class for room returns a description and the status of the player can be found throughout the game. As for pickupitem... Don't you want to just pick up one item at a time from the room? Also why does the minotaur just stand there and take your attacks? What do you have against the noble and majestic creature that is the minotaur that you should so callously transform it into a punching bag to be pommelled until death? Also you might want a maximum health and a finite number of healing supplements otherwise I _am_ going to play this game healing my character until i max out space on my computer to contain all the digits that'll be my player health.

Hello, I really do like your ideas of the player input functions. This would make the game feel a lot more complete. The minotaur now is not a punching bag and the idea of finite health potions did come to mind. I love these ideas and definitely will look into them in the future. Currently the way the game works does the trick and meets the requirements, however I do love these points and I believe they would polish the game.
Thank you very much for these reviews and taking the time to give me some helpful feedback, I appreciate it a lot.

eoin2406 commented:
> Error Handling: Your code won't throw up exceptions and this is good. However you might want to consider using .ToLower() and Trim() functions for player input. These will ensure that if a player enters ' a' [note the white space] or 'A' to attack, then these inputs will also be accepted. You've used conditionals which is fine. I haven't seen any System.Diagnostics.Debug.Assert() used to check exceptions won't be thrown preemptively. And there are no try and catch statements. As you continue to develop your code, you might want to consider deploying these techniques. Try and catch are especially useful when your code gets more complex and you want to use a function that might only apply some of the time and throw exceptions on others.
> 
> Encapsulation and Abstraction: You've used the internal scope and private fields which is great. I noticed that you might have some confusion as to what classes are for and how they are distinct from objects. This is because [and i mentioned above] your classes' functions often return stock responses to the console that isn't necessarily going to apply to every object your instantiate with said class. Simple interpolation is test way to get around this in most cases. There are no abstract classes from what I can tell and you have yet to deploy inheritance.
> 
> Methods and Constructors: You've made the constructors well and differently from how we were taught in class which I found interesting. Each class for room returns a description and the status of the player can be found throughout the game. As for pickupitem... Don't you want to just pick up one item at a time from the room? Also why does the minotaur just stand there and take your attacks? What do you have against the noble and majestic creature that is the minotaur that you should so callously transform it into a punching bag to be pommelled until death? Also you might want a maximum health and a finite number of healing supplements otherwise I _am_ going to play this game healing my character until i max out space on my computer to contain all the digits that'll be my player health.

Hello again, I actually went back and used the ToLower() and Trim() functions. They make the game a lot nicer for user input and make the chance of errors a lot less common.
Thank you so much for this recommendation, I really like this and appreciate this feedback a lot.

eoin2406 committed:
Charles' and Ben's Suggestions and some other tweaks

After acting upon Charles' reviews and suggestions, I focused on Ben's. I really liked the idea of adding the ToLower() and Trim() functions for user input. This allowed the game to feel more complete and more playable for users. I also added comments throughout the code, for presentation purposes, as well as allowing it to be more understandable as a whole.
